### DISCUSSION TITLE
Vercel dev CLI using old typescript version

### DISCUSSION THREAD

**Discussion by nipunn1313** [vercel/vercel#8558] [Help]
Hi,

In my project, I'm running vercel dev from the CLI to run a serverless function. When my serverless function is invoked, it appears to run typescript on my serverless function using an old version of typescript.

Example error:
```
node_modules/convex/dist/esm/browser/browser.d.ts:174:39 - error TS2304: Cannot find name 'Awaited'.
```

174     (...args: Parameters<F>): Promise<Awaited<ReturnType<F>>>;
Awaited was introduced in typescript 4.5.

My local package.json references typescript 4.7 and running `npx tsc` locally passes.

EDIT:
While debugging I found this:
```
➜  4b-external-functions-vercel git:(main) ✗ for t in `find . -name tsc`; do echo $t ; $t --version; done
./node_modules/.bin/tsc
Version 4.7.4
./node_modules/typescript/bin/tsc
Version 4.7.4
./node_modules/@vercel/node/node_modules/.bin/tsc
Version 4.3.4
./node_modules/@vercel/node/node_modules/typescript/bin/tsc
Version 4.3.4
```

So I'm suspicious that `@vercel/node` is shipping with an old version of typescript and then choosing that version over the one I have locally installed from my package.json.

**Comment by swarnava** [comment]
Have you tried removing `package-lock.json` and then deploy?

**Reply by nipunn1313** [reply]
I have tried this. Still gets same results.
I believe it's likely that `@vercel/node` is shipping with an old version of typescript.

**Reply by swarnava** [reply]
Could you share a [minimal repository](https://stackoverflow.com/help/minimal-reproducible-example) which I can reproduce?

**Reply by nipunn1313** [reply]
Here is a minimal repro
https://github.com/nipunn1313/repro_vercel_ts_issue

Here's a screenshot of me running the repro
<img width="478" alt="image" src="https://user-images.githubusercontent.com/1300387/190921291-95342191-618b-464e-b462-9cf46cc15c90.png">

**Reply by vinh-perfectstorm** [reply]
Having a similar problem using vercel serverless and sveltekit which uses compiler option “preserveValueImports”

**Reply by amyegan** [reply]
Thank you for sharing the minimal repro @nipunn1313. This is an interesting problem. The only time I can see it happen is on a local `vc dev` run. It does not seem to be a problem for preview and production deployments based on what I am seeing.

You can see in this repo that `@vercel/node` does have `typescript@4.3.4` as a dependency. https://github.com/vercel/vercel/blob/053c185481b1fa0f980e3f54c9ecbeda53e7d72b/packages/node/package.json#L42

I'm not sure what might break if we update that package to a more recent version. I opened a new issue to look at updating the dependency (https://github.com/vercel/vercel/issues/8680). The repo is open source, so feel free to experiment with it yourself as well!

**Reply by nipunn1313** [reply]
Hi Amy! Thanks for taking a look.

I don't think upgrading `typescript@4.3.4` is the ideal fix (though I wouldn't be against it independently).

IMO, Vercel should either
- Run the version of typescript that is in the package.json of MY project. Not the one bundled with vercel.
- Stop running typescript dynamically on api/ functions at the time they are called

open to being convinced otherwise.
Upgrading typescript still means there is a discrepancy between the typescript version specified by project owner in package.json and the one being run by vercel.

**Reply by amyegan** [reply]
Thanks for clarifying! I updated the issue summary

**Reply by vxna** [reply]
<img width="661" alt="CleanShot 2023-01-06 at 18 26 42@2x" src="https://user-images.githubusercontent.com/30002276/211012537-d5cec4c4-dc74-4669-b40b-b28809f22fdf.png">
<img width="570" alt="CleanShot 2023-01-06 at 18 27 11@2x" src="https://user-images.githubusercontent.com/30002276/211012597-90e8e92d-1c6e-437a-83f5-4b5133fdf27d.png">

Having the same problem using the `got` package with `vc dev`. 
@amyegan I think it's worth converting this discussion thread into an issue 👍 

ref: https://github.com/sindresorhus/got/issues/2164#issuecomment-1284315448

**Reply by amyegan** [reply]
Thanks for sharing this info @vxna. You can find the issue created from the original post here https://github.com/vercel/vercel/issues/8680

**Reply by martin-dimi** [reply]
@vxna I'm having the same issue. Did you manage to fix this?

**Comment by netgfx** [comment]
I think my issue is related to this as well I get `/bin/sh: tsc: command not found` when trying to build types of my repository. The failing command is: `"build:types": "tsc --declaration --emitDeclarationOnly",` 
What can be done about this? I have included `typescript` as a devDependency `4.9.2`

**Reply by AnimeshTimsina** [reply]
how did you solve this?

**Reply by netgfx** [reply]
I didn't. I set up vercel to pull from repository instead and then it wasn't having any issues. The issue was caused when I tried to deploy from Codesanbox through vercel
